vllm/distributed/kv_transfer/kv_connector/v1/offloading_connector.py

```diff
@@ -0,0 +1,469 @@
+
+        block_hashes = [
+            blk_hash.hash_value
+            for blk_hash in request.block_hashes[self.block_size_factor -
+                                                 1::self.block_size_factor]
+        ]
+        assert len(block_hashes) == num_blocks
+
+        self.manager.touch(block_hashes)
+
+        full_block_tokens = self.offloaded_block_size * num_blocks
+        if full_block_tokens - num_computed_tokens < self.offloaded_block_size:
+            # we can load less than a block, skip
+            return 0, False
+
+        start_block_idx = num_computed_tokens // self.offloaded_block_size
+        hits = self.manager.lookup(block_hashes[start_block_idx:])
+        if hits == 0:
+            return 0, False
+
+        num_hit_tokens = (self.offloaded_block_size *
+                          (start_block_idx + hits) - num_computed_tokens)
+        logger.info(
+            "Request %s hit %s offloaded tokens after %s GPU hit tokens",
+            request.request_id,
+            num_hit_tokens,
+            num_computed_tokens,
+        )
+        if num_hit_tokens < self.offloaded_block_size:
+            return 0, False
+
+        return num_hit_tokens, True
+
+    def update_state_after_alloc(self, request: Request, blocks: KVCacheBlocks,
+                                 num_external_tokens: int):
+        self._requests[request.request_id] = request
+        # the block ids are updated in _get_reqs_to_store
+        self._request_block_ids[request.request_id] = []
+
+        if num_external_tokens == 0:
+            return
+
+        block_groups = blocks.get_block_ids()
+        assert len(block_groups) == 1, "Only one group is supported"
+        block_ids = block_groups[0]
+
+        num_computed_gpu_blocks = sum(block.block_hash is not None
+                                      for block in blocks.blocks[0])
+        num_computed_tokens = num_computed_gpu_blocks * self.gpu_block_size
+        full_block_tokens = num_computed_tokens + num_external_tokens
+        assert full_block_tokens % self.offloaded_block_size == 0
+
+        num_pending_gpu_blocks = len(block_ids) - num_computed_gpu_blocks
+        assert (num_external_tokens == num_pending_gpu_blocks *
+                self.gpu_block_size)
+
+        start_block_idx = num_computed_tokens // self.offloaded_block_size
+        num_blocks = full_block_tokens // self.offloaded_block_size
+
+        block_hashes = [
+            blk_hash.hash_value
+            for blk_hash in request.block_hashes[self.block_size_factor -
+                                                 1::self.block_size_factor]
+        ]
+        assert len(block_hashes) >= num_blocks
+
+        block_hashes = block_hashes[start_block_idx:num_blocks]
+
+        src_specs = self.manager.prepare_load(block_hashes)
+        dst_specs = [
+            GPULoadStoreSpec(gpu_block_id)
+            for gpu_block_id in block_ids[num_computed_gpu_blocks:]
```

Comment:
use islice?

Reply:
I may be misunderstanding but this comprehension is creating a list of `GPULoadStoreSpec` objects, which can't be used to create a tensor directly?